From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jorge Betancourt <jmbetancourt@google.com>
Date: Thu, 21 May 2026 11:25:35 -0400
Subject: [M149] Clamp displacement values before offsetting sampling
 coordinates

Original change's description:
> Clamp displacement values before offsetting sampling coordinates
>
> Bug: b/514082801
> Change-Id: I4662eb2d969bc6003cd7178306e82b6d3580758a
> Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1239498
> Auto-Submit: Jorge Betancourt <jmbetancourt@google.com>
> Reviewed-by: Michael Ludwig <michaelludwig@google.com>

(cherry picked from commit bd3d88bbbcfef52f755833215bcc1d1ce695bc26)

Bug: 515954609,b/514082801
Change-Id: I4662eb2d969bc6003cd7178306e82b6d3580758a
Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1247396
Auto-Submit: Chrome Cherry Picker <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>

diff --git a/src/sksl/generated/sksl_rt_shader.minified.sksl b/src/sksl/generated/sksl_rt_shader.minified.sksl
index c15e6a11f9ceb528d514e8a8ea10f803b56db169..50a39b747dbb976b40578af52bac3cfe9765f1a5 100644
--- a/src/sksl/generated/sksl_rt_shader.minified.sksl
+++ b/src/sksl/generated/sksl_rt_shader.minified.sksl
@@ -3,44 +3,44 @@ static constexpr char SKSL_MINIFIED_sksl_rt_shader[] =
 ",0.,0.,saturate(dot(half3(.2126,.7152,.0722),a)));}half4 sk_decal(shader a,"
 "float2 b,float4 c){half4 e=half4(c-b.xyxy)*half4(-1.,-1.,1.,1.);e=saturate("
 "e+.5);return(((e.x*e.y)*e.z)*e.w)*a.eval(b);}half4 sk_displacement(shader a"
-",shader b,float2 c,half2 d,half4 e,half4 f){half4 g=unpremul(a.eval(c));half2"
-" h=half2(dot(g,e),dot(g,f));h=d*(h-.5);return b.eval(c+float2(h));}half4 sk_magnifier"
-"(shader a,float2 b,float4 c,float4 d,float2 e){float2 f=d.xy+d.zw*b;float2 g"
-"=min(b-c.xy,c.zw-b)*e;float h=all(lessThan(g,2..xx))?2.-length(2.-g):min(g."
-"x,g.y);h=saturate(h);return a.eval(mix(b,f,h*h));}$pure half3 $b(half3 a){half"
-" b=max(max(a.x,a.y),a.z);half e=min(min(a.x,a.y),a.z);half f=b-e;half g=1./"
-"f;half i=half(a.y<a.z?6.:0.);half j=.166666672*(b==e?0.:(a.x>=a.y&&a.x>=a.z"
-"?g*(a.y-a.z)+i:(a.y>=a.z?g*(a.z-a.x)+2.:g*(a.x-a.y)+4.)));half k=b+e;half m"
-"=k*.5;half n=b==e?0.:f/(m>.5?2.-k:k);return half3(j,n,m);}half3 sk_high_contrast"
-"(half3 a,half b,half c,half d){if(b==1.)a=dot(half3(.2126,.7152,.0722),a).xxx"
-";if(c==1.)a=1.-a;else if(c==2.){a=$b(a);a.z=1.-a.z;a=$hsl_to_rgb(a);}return"
-" saturate(mix(half3(.5),a,d));}$pure half3 $c(half3 a,half3 b,half3 c,half d"
-"){const half3 e=half3(.25,.5,.25);half3 f=half3(a.x,b.x,c.x);half3 g=half3("
-"a.z,b.z,c.z);half h=dot(e,c)-dot(e,a);half i=dot(e,g)-dot(e,f);return normalize"
-"(half3(d*half2(h,i),1.));}half4 sk_normal(shader a,float2 b,float4 c,half d"
-"){half3 e=half3(a.eval(clamp(b+ -1..xx,c.xy,c.zw)).w,a.eval(clamp(b+float2("
-"-1.,0.),c.xy,c.zw)).w,a.eval(clamp(b+float2(-1.,1.),c.xy,c.zw)).w);half3 f="
-"half3(a.eval(clamp(b+float2(0.,-1.),c.xy,c.zw)).w,a.eval(clamp(b,c.xy,c.zw)"
-").w,a.eval(clamp(b+float2(0.,1.),c.xy,c.zw)).w);half3 g=half3(a.eval(clamp("
-"b+float2(1.,-1.),c.xy,c.zw)).w,a.eval(clamp(b+float2(1.,0.),c.xy,c.zw)).w,a"
-".eval(clamp(b+1..xx,c.xy,c.zw)).w);half h=f.y;return half4($c(e,f,g,d),h);}"
-"$pure half3 $d(half a,half3 b,half3 c,half3 d){return a>=0.?normalize(b-d):"
-"c;}$pure half $e(half3 a,half3 b,half c,half d){half g=-dot(b,a);if(g<c)return"
-" 0.;else{half h=pow(g,d);return g<c+.016?(h*(g-c))*62.4999962:h;}}$pure half4"
-" $f(half3 a,half b,half c,half d,half3 e,half3 f,half3 g,half h,half i){if("
-"d>0.)a*=$e(f,g,h,i);if(c==0.){half j=dot(e,g);a=saturate(j*a);return half4("
-"a,1.);}else if(c==1.){half3 j=normalize(g+half3(0.,0.,1.));half k=pow(dot(e"
-",j),b);a=saturate(k*a);return half4(a,max(max(a.x,a.y),a.z));}else{half j=("
-"2.*dot(e,g)-g.z)*g.z;half k=pow(j,b);a=saturate(k*a);return half4(a,max(max"
-"(a.x,a.y),a.z));}}half4 sk_lighting(shader a,float2 b,half c,half d,half e,"
-"half f,half3 g,half h,half3 i,half j,half3 k){half4 l=a.eval(b);half3 m=$d("
-"f,g,i,half3(half2(b),c*l.w));return $f(k,d,e,f,l.xyz,i,m,j,h);}half4 sk_arithmetic_blend"
-"(half4 a,half4 b,half4 c,half d){half4 e=saturate((((c.x*a)*b+c.y*a)+c.z*b)"
-"+c.w);e.xyz=min(e.xyz,max(e.w,d));return e;}half4 sk_sparse_morphology(shader"
-" a,float2 b,half2 c,half d){half4 e=max(d*a.eval(b+float2(c)),d*a.eval(b-float2"
-"(c)));return d*e;}half4 sk_linear_morphology(shader a,float2 b,half2 c,half"
-" d,int e){half4 g=d*a.eval(b);half2 h=c;for(int j=1;j<=14;++j){if(j>e)break"
-";g=max(g,max(d*a.eval(b+float2(h)),d*a.eval(b-float2(h))));h+=c;}return d*g"
-";}half4 sk_overdraw(half a,half4 b,half4 c,half4 d,half4 e,half4 f,half4 g)"
-"{return a<.00196078443?b:(a<.005882353?c:(a<.009803922?d:(a<.01372549?e:(a<"
-".01764706?f:g))));}";
+",shader b,float2 c,half2 d,half4 e,half4 f){half4 g=saturate(unpremul(a.eval"
+"(c)));half2 h=half2(dot(g,e),dot(g,f));h=d*(h-.5);return b.eval(c+float2(h)"
+");}half4 sk_magnifier(shader a,float2 b,float4 c,float4 d,float2 e){float2 f"
+"=d.xy+d.zw*b;float2 g=min(b-c.xy,c.zw-b)*e;float h=all(lessThan(g,2..xx))?2."
+"-length(2.-g):min(g.x,g.y);h=saturate(h);return a.eval(mix(b,f,h*h));}$pure"
+" half3 $b(half3 a){half b=max(max(a.x,a.y),a.z);half e=min(min(a.x,a.y),a.z"
+");half f=b-e;half g=1./f;half i=half(a.y<a.z?6.:0.);half j=.166666672*(b==e"
+"?0.:(a.x>=a.y&&a.x>=a.z?g*(a.y-a.z)+i:(a.y>=a.z?g*(a.z-a.x)+2.:g*(a.x-a.y)+"
+"4.)));half k=b+e;half m=k*.5;half n=b==e?0.:f/(m>.5?2.-k:k);return half3(j,"
+"n,m);}half3 sk_high_contrast(half3 a,half b,half c,half d){if(b==1.)a=dot(half3"
+"(.2126,.7152,.0722),a).xxx;if(c==1.)a=1.-a;else if(c==2.){a=$b(a);a.z=1.-a."
+"z;a=$hsl_to_rgb(a);}return saturate(mix(half3(.5),a,d));}$pure half3 $c(half3"
+" a,half3 b,half3 c,half d){const half3 e=half3(.25,.5,.25);half3 f=half3(a."
+"x,b.x,c.x);half3 g=half3(a.z,b.z,c.z);half h=dot(e,c)-dot(e,a);half i=dot(e"
+",g)-dot(e,f);return normalize(half3(d*half2(h,i),1.));}half4 sk_normal(shader"
+" a,float2 b,float4 c,half d){half3 e=half3(a.eval(clamp(b+ -1..xx,c.xy,c.zw"
+")).w,a.eval(clamp(b+float2(-1.,0.),c.xy,c.zw)).w,a.eval(clamp(b+float2(-1.,"
+"1.),c.xy,c.zw)).w);half3 f=half3(a.eval(clamp(b+float2(0.,-1.),c.xy,c.zw))."
+"w,a.eval(clamp(b,c.xy,c.zw)).w,a.eval(clamp(b+float2(0.,1.),c.xy,c.zw)).w);"
+"half3 g=half3(a.eval(clamp(b+float2(1.,-1.),c.xy,c.zw)).w,a.eval(clamp(b+float2"
+"(1.,0.),c.xy,c.zw)).w,a.eval(clamp(b+1..xx,c.xy,c.zw)).w);half h=f.y;return"
+" half4($c(e,f,g,d),h);}$pure half3 $d(half a,half3 b,half3 c,half3 d){return"
+" a>=0.?normalize(b-d):c;}$pure half $e(half3 a,half3 b,half c,half d){half g"
+"=-dot(b,a);if(g<c)return 0.;else{half h=pow(g,d);return g<c+.016?(h*(g-c))*"
+"62.4999962:h;}}$pure half4 $f(half3 a,half b,half c,half d,half3 e,half3 f,"
+"half3 g,half h,half i){if(d>0.)a*=$e(f,g,h,i);if(c==0.){half j=dot(e,g);a=saturate"
+"(j*a);return half4(a,1.);}else if(c==1.){half3 j=normalize(g+half3(0.,0.,1."
+"));half k=pow(dot(e,j),b);a=saturate(k*a);return half4(a,max(max(a.x,a.y),a"
+".z));}else{half j=(2.*dot(e,g)-g.z)*g.z;half k=pow(j,b);a=saturate(k*a);return"
+" half4(a,max(max(a.x,a.y),a.z));}}half4 sk_lighting(shader a,float2 b,half c"
+",half d,half e,half f,half3 g,half h,half3 i,half j,half3 k){half4 l=a.eval"
+"(b);half3 m=$d(f,g,i,half3(half2(b),c*l.w));return $f(k,d,e,f,l.xyz,i,m,j,h"
+");}half4 sk_arithmetic_blend(half4 a,half4 b,half4 c,half d){half4 e=saturate"
+"((((c.x*a)*b+c.y*a)+c.z*b)+c.w);e.xyz=min(e.xyz,max(e.w,d));return e;}half4"
+" sk_sparse_morphology(shader a,float2 b,half2 c,half d){half4 e=max(d*a.eval"
+"(b+float2(c)),d*a.eval(b-float2(c)));return d*e;}half4 sk_linear_morphology"
+"(shader a,float2 b,half2 c,half d,int e){half4 g=d*a.eval(b);half2 h=c;for("
+"int j=1;j<=14;++j){if(j>e)break;g=max(g,max(d*a.eval(b+float2(h)),d*a.eval("
+"b-float2(h))));h+=c;}return d*g;}half4 sk_overdraw(half a,half4 b,half4 c,half4"
+" d,half4 e,half4 f,half4 g){return a<.00196078443?b:(a<.005882353?c:(a<.009803922"
+"?d:(a<.01372549?e:(a<.01764706?f:g))));}";
diff --git a/src/sksl/generated/sksl_rt_shader.unoptimized.sksl b/src/sksl/generated/sksl_rt_shader.unoptimized.sksl
index b0c764eee9128fa34a575fee5d100d481873f1e6..b3c09b83f7aa0ca76f694138d804b0905f9443b8 100644
--- a/src/sksl/generated/sksl_rt_shader.unoptimized.sksl
+++ b/src/sksl/generated/sksl_rt_shader.unoptimized.sksl
@@ -4,72 +4,73 @@ static constexpr char SKSL_MINIFIED_sksl_rt_shader[] =
 " image,float2 coord,float4 decalBounds){half4 d=half4(decalBounds-coord.xyxy"
 ")*half4(-1.,-1.,1.,1.);d=saturate(d+.5);return(((d.x*d.y)*d.z)*d.w)*image.eval"
 "(coord);}half4 sk_displacement(shader displMap,shader colorMap,float2 coord"
-",half2 scale,half4 xSelect,half4 ySelect){half4 displColor=unpremul(displMap"
-".eval(coord));half2 displ=half2(dot(displColor,xSelect),dot(displColor,ySelect"
-"));displ=scale*(displ-.5);return colorMap.eval(coord+float2(displ));}half4 sk_magnifier"
-"(shader src,float2 coord,float4 lensBounds,float4 zoomXform,float2 invInset"
-"){float2 zoomCoord=zoomXform.xy+zoomXform.zw*coord;float2 edgeInset=min(coord"
-"-lensBounds.xy,lensBounds.zw-coord)*invInset;float weight=all(lessThan(edgeInset"
-",2..xx))?2.-length(2.-edgeInset):min(edgeInset.x,edgeInset.y);weight=saturate"
-"(weight);return src.eval(mix(coord,zoomCoord,weight*weight));}$pure half3 $high_contrast_rgb_to_hsl"
-"(half3 c){half mx=max(max(c.x,c.y),c.z);half mn=min(min(c.x,c.y),c.z);half d"
-"=mx-mn;half invd=1./d;half g_lt_b=half(c.y<c.z?6.:0.);half h=.166666672*(mx"
-"==mn?0.:(c.x>=c.y&&c.x>=c.z?invd*(c.y-c.z)+g_lt_b:(c.y>=c.z?invd*(c.z-c.x)+"
-"2.:invd*(c.x-c.y)+4.)));half sum=mx+mn;half l=sum*.5;half s=mx==mn?0.:d/(l>"
-".5?2.-sum:sum);return half3(h,s,l);}half3 sk_high_contrast(half3 color,half"
-" grayscale,half invertStyle,half contrast){if(grayscale==1.)color=dot(half3"
-"(.2126,.7152,.0722),color).xxx;if(invertStyle==1.)color=1.-color;else if(invertStyle"
-"==2.){color=$high_contrast_rgb_to_hsl(color);color.z=1.-color.z;color=$hsl_to_rgb"
-"(color);}return saturate(mix(half3(.5),color,contrast));}$pure half3 $normal_filter"
-"(half3 alphaC0,half3 alphaC1,half3 alphaC2,half negSurfaceDepth){const half3"
-" kSobel=half3(.25,.5,.25);half3 alphaR0=half3(alphaC0.x,alphaC1.x,alphaC2.x"
-");half3 alphaR2=half3(alphaC0.z,alphaC1.z,alphaC2.z);half nx=dot(kSobel,alphaC2"
-")-dot(kSobel,alphaC0);half ny=dot(kSobel,alphaR2)-dot(kSobel,alphaR0);return"
-" normalize(half3(negSurfaceDepth*half2(nx,ny),1.));}half4 sk_normal(shader alphaMap"
-",float2 coord,float4 edgeBounds,half negSurfaceDepth){half3 alphaC0=half3(alphaMap"
-".eval(clamp(coord+ -1..xx,edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval(clamp"
-"(coord+float2(-1.,0.),edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval(clamp(coord"
-"+float2(-1.,1.),edgeBounds.xy,edgeBounds.zw)).w);half3 alphaC1=half3(alphaMap"
-".eval(clamp(coord+float2(0.,-1.),edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval"
-"(clamp(coord,edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval(clamp(coord+float2"
-"(0.,1.),edgeBounds.xy,edgeBounds.zw)).w);half3 alphaC2=half3(alphaMap.eval("
-"clamp(coord+float2(1.,-1.),edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval(clamp"
-"(coord+float2(1.,0.),edgeBounds.xy,edgeBounds.zw)).w,alphaMap.eval(clamp(coord"
-"+1..xx,edgeBounds.xy,edgeBounds.zw)).w);half mainAlpha=alphaC1.y;return half4"
-"($normal_filter(alphaC0,alphaC1,alphaC2,negSurfaceDepth),mainAlpha);}$pure half3"
-" $surface_to_light(half lightType,half3 lightPos,half3 lightDir,half3 coord"
-"){return lightType>=0.?normalize(lightPos-coord):lightDir;}$pure half $spotlight_scale"
-"(half3 lightDir,half3 surfaceToLight,half cosCutoffAngle,half spotFalloff){"
-"const half kConeAAThreshold=.016;const half kConeScale=62.4999962;half cosAngle"
-"=-dot(surfaceToLight,lightDir);if(cosAngle<cosCutoffAngle)return 0.;else{half"
-" scale=pow(cosAngle,spotFalloff);return cosAngle<cosCutoffAngle+kConeAAThreshold"
-"?(scale*(cosAngle-cosCutoffAngle))*kConeScale:scale;}}$pure half4 $compute_lighting"
-"(half3 color,half shininess,half materialType,half lightType,half3 normal,half3"
-" lightDir,half3 surfaceToLight,half cosCutoffAngle,half spotFalloff){if(lightType"
-">0.)color*=$spotlight_scale(lightDir,surfaceToLight,cosCutoffAngle,spotFalloff"
-");if(materialType==0.){half coeff=dot(normal,surfaceToLight);color=saturate"
-"(coeff*color);return half4(color,1.);}else if(materialType==1.){half3 halfDir"
-"=normalize(surfaceToLight+half3(0.,0.,1.));half coeff=pow(dot(normal,halfDir"
-"),shininess);color=saturate(coeff*color);return half4(color,max(max(color.x"
-",color.y),color.z));}else{half hilite=(2.*dot(normal,surfaceToLight)-surfaceToLight"
-".z)*surfaceToLight.z;half coeff=pow(hilite,shininess);color=saturate(coeff*"
-"color);return half4(color,max(max(color.x,color.y),color.z));}}half4 sk_lighting"
-"(shader normalMap,float2 coord,half depth,half shininess,half materialType,"
-"half lightType,half3 lightPos,half spotFalloff,half3 lightDir,half cosCutoffAngle"
-",half3 lightColor){half4 normalAndA=normalMap.eval(coord);half3 surfaceToLight"
-"=$surface_to_light(lightType,lightPos,lightDir,half3(half2(coord),depth*normalAndA"
-".w));return $compute_lighting(lightColor,shininess,materialType,lightType,normalAndA"
-".xyz,lightDir,surfaceToLight,cosCutoffAngle,spotFalloff);}half4 sk_arithmetic_blend"
-"(half4 src,half4 dst,half4 k,half pmClamp){half4 color=saturate((((k.x*src)"
-"*dst+k.y*src)+k.z*dst)+k.w);color.xyz=min(color.xyz,max(color.w,pmClamp));return"
-" color;}half4 sk_sparse_morphology(shader child,float2 coord,half2 offset,half"
-" flip){half4 aggregate=max(flip*child.eval(coord+float2(offset)),flip*child"
-".eval(coord-float2(offset)));return flip*aggregate;}half4 sk_linear_morphology"
-"(shader child,float2 coord,half2 offset,half flip,int radius){const int kMaxLinearRadius"
-"=14;half4 aggregate=flip*child.eval(coord);half2 delta=offset;for(int i=1;i"
-"<=kMaxLinearRadius;++i){if(i>radius)break;aggregate=max(aggregate,max(flip*"
-"child.eval(coord+float2(delta)),flip*child.eval(coord-float2(delta))));delta"
-"+=offset;}return flip*aggregate;}half4 sk_overdraw(half alpha,half4 color0,"
-"half4 color1,half4 color2,half4 color3,half4 color4,half4 color5){return alpha"
-"<.00196078443?color0:(alpha<.005882353?color1:(alpha<.009803922?color2:(alpha"
-"<.01372549?color3:(alpha<.01764706?color4:color5))));}";
+",half2 scale,half4 xSelect,half4 ySelect){half4 displColor=saturate(unpremul"
+"(displMap.eval(coord)));half2 displ=half2(dot(displColor,xSelect),dot(displColor"
+",ySelect));displ=scale*(displ-.5);return colorMap.eval(coord+float2(displ))"
+";}half4 sk_magnifier(shader src,float2 coord,float4 lensBounds,float4 zoomXform"
+",float2 invInset){float2 zoomCoord=zoomXform.xy+zoomXform.zw*coord;float2 edgeInset"
+"=min(coord-lensBounds.xy,lensBounds.zw-coord)*invInset;float weight=all(lessThan"
+"(edgeInset,2..xx))?2.-length(2.-edgeInset):min(edgeInset.x,edgeInset.y);weight"
+"=saturate(weight);return src.eval(mix(coord,zoomCoord,weight*weight));}$pure"
+" half3 $high_contrast_rgb_to_hsl(half3 c){half mx=max(max(c.x,c.y),c.z);half"
+" mn=min(min(c.x,c.y),c.z);half d=mx-mn;half invd=1./d;half g_lt_b=half(c.y<"
+"c.z?6.:0.);half h=.166666672*(mx==mn?0.:(c.x>=c.y&&c.x>=c.z?invd*(c.y-c.z)+"
+"g_lt_b:(c.y>=c.z?invd*(c.z-c.x)+2.:invd*(c.x-c.y)+4.)));half sum=mx+mn;half"
+" l=sum*.5;half s=mx==mn?0.:d/(l>.5?2.-sum:sum);return half3(h,s,l);}half3 sk_high_contrast"
+"(half3 color,half grayscale,half invertStyle,half contrast){if(grayscale==1."
+")color=dot(half3(.2126,.7152,.0722),color).xxx;if(invertStyle==1.)color=1.-"
+"color;else if(invertStyle==2.){color=$high_contrast_rgb_to_hsl(color);color"
+".z=1.-color.z;color=$hsl_to_rgb(color);}return saturate(mix(half3(.5),color"
+",contrast));}$pure half3 $normal_filter(half3 alphaC0,half3 alphaC1,half3 alphaC2"
+",half negSurfaceDepth){const half3 kSobel=half3(.25,.5,.25);half3 alphaR0=half3"
+"(alphaC0.x,alphaC1.x,alphaC2.x);half3 alphaR2=half3(alphaC0.z,alphaC1.z,alphaC2"
+".z);half nx=dot(kSobel,alphaC2)-dot(kSobel,alphaC0);half ny=dot(kSobel,alphaR2"
+")-dot(kSobel,alphaR0);return normalize(half3(negSurfaceDepth*half2(nx,ny),1."
+"));}half4 sk_normal(shader alphaMap,float2 coord,float4 edgeBounds,half negSurfaceDepth"
+"){half3 alphaC0=half3(alphaMap.eval(clamp(coord+ -1..xx,edgeBounds.xy,edgeBounds"
+".zw)).w,alphaMap.eval(clamp(coord+float2(-1.,0.),edgeBounds.xy,edgeBounds.zw"
+")).w,alphaMap.eval(clamp(coord+float2(-1.,1.),edgeBounds.xy,edgeBounds.zw))"
+".w);half3 alphaC1=half3(alphaMap.eval(clamp(coord+float2(0.,-1.),edgeBounds"
+".xy,edgeBounds.zw)).w,alphaMap.eval(clamp(coord,edgeBounds.xy,edgeBounds.zw"
+")).w,alphaMap.eval(clamp(coord+float2(0.,1.),edgeBounds.xy,edgeBounds.zw))."
+"w);half3 alphaC2=half3(alphaMap.eval(clamp(coord+float2(1.,-1.),edgeBounds."
+"xy,edgeBounds.zw)).w,alphaMap.eval(clamp(coord+float2(1.,0.),edgeBounds.xy,"
+"edgeBounds.zw)).w,alphaMap.eval(clamp(coord+1..xx,edgeBounds.xy,edgeBounds."
+"zw)).w);half mainAlpha=alphaC1.y;return half4($normal_filter(alphaC0,alphaC1"
+",alphaC2,negSurfaceDepth),mainAlpha);}$pure half3 $surface_to_light(half lightType"
+",half3 lightPos,half3 lightDir,half3 coord){return lightType>=0.?normalize("
+"lightPos-coord):lightDir;}$pure half $spotlight_scale(half3 lightDir,half3 surfaceToLight"
+",half cosCutoffAngle,half spotFalloff){const half kConeAAThreshold=.016;const"
+" half kConeScale=62.4999962;half cosAngle=-dot(surfaceToLight,lightDir);if("
+"cosAngle<cosCutoffAngle)return 0.;else{half scale=pow(cosAngle,spotFalloff)"
+";return cosAngle<cosCutoffAngle+kConeAAThreshold?(scale*(cosAngle-cosCutoffAngle"
+"))*kConeScale:scale;}}$pure half4 $compute_lighting(half3 color,half shininess"
+",half materialType,half lightType,half3 normal,half3 lightDir,half3 surfaceToLight"
+",half cosCutoffAngle,half spotFalloff){if(lightType>0.)color*=$spotlight_scale"
+"(lightDir,surfaceToLight,cosCutoffAngle,spotFalloff);if(materialType==0.){half"
+" coeff=dot(normal,surfaceToLight);color=saturate(coeff*color);return half4("
+"color,1.);}else if(materialType==1.){half3 halfDir=normalize(surfaceToLight"
+"+half3(0.,0.,1.));half coeff=pow(dot(normal,halfDir),shininess);color=saturate"
+"(coeff*color);return half4(color,max(max(color.x,color.y),color.z));}else{half"
+" hilite=(2.*dot(normal,surfaceToLight)-surfaceToLight.z)*surfaceToLight.z;half"
+" coeff=pow(hilite,shininess);color=saturate(coeff*color);return half4(color"
+",max(max(color.x,color.y),color.z));}}half4 sk_lighting(shader normalMap,float2"
+" coord,half depth,half shininess,half materialType,half lightType,half3 lightPos"
+",half spotFalloff,half3 lightDir,half cosCutoffAngle,half3 lightColor){half4"
+" normalAndA=normalMap.eval(coord);half3 surfaceToLight=$surface_to_light(lightType"
+",lightPos,lightDir,half3(half2(coord),depth*normalAndA.w));return $compute_lighting"
+"(lightColor,shininess,materialType,lightType,normalAndA.xyz,lightDir,surfaceToLight"
+",cosCutoffAngle,spotFalloff);}half4 sk_arithmetic_blend(half4 src,half4 dst"
+",half4 k,half pmClamp){half4 color=saturate((((k.x*src)*dst+k.y*src)+k.z*dst"
+")+k.w);color.xyz=min(color.xyz,max(color.w,pmClamp));return color;}half4 sk_sparse_morphology"
+"(shader child,float2 coord,half2 offset,half flip){half4 aggregate=max(flip"
+"*child.eval(coord+float2(offset)),flip*child.eval(coord-float2(offset)));return"
+" flip*aggregate;}half4 sk_linear_morphology(shader child,float2 coord,half2"
+" offset,half flip,int radius){const int kMaxLinearRadius=14;half4 aggregate"
+"=flip*child.eval(coord);half2 delta=offset;for(int i=1;i<=kMaxLinearRadius;"
+"++i){if(i>radius)break;aggregate=max(aggregate,max(flip*child.eval(coord+float2"
+"(delta)),flip*child.eval(coord-float2(delta))));delta+=offset;}return flip*"
+"aggregate;}half4 sk_overdraw(half alpha,half4 color0,half4 color1,half4 color2"
+",half4 color3,half4 color4,half4 color5){return alpha<.00196078443?color0:("
+"alpha<.005882353?color1:(alpha<.009803922?color2:(alpha<.01372549?color3:(alpha"
+"<.01764706?color4:color5))));}";
diff --git a/src/sksl/sksl_rt_shader.sksl b/src/sksl/sksl_rt_shader.sksl
index 811deb3b9625021876cceea87de28c094b993407..03378cb0bc748939a1813938643202e8c3d180a9 100644
--- a/src/sksl/sksl_rt_shader.sksl
+++ b/src/sksl/sksl_rt_shader.sksl
@@ -22,7 +22,7 @@ half4 sk_displacement(shader displMap,
                       half2 scale,
                       half4 xSelect,  // Only one of RGBA will be 1, the rest are 0
                       half4 ySelect) {
-    half4 displColor = unpremul(displMap.eval(coord));
+    half4 displColor = saturate(unpremul(displMap.eval(coord)));
     half2 displ = half2(dot(displColor, xSelect), dot(displColor, ySelect));
     displ = scale * (displ - 0.5);
     return colorMap.eval(coord + displ);
